feat: new react template based code generation

anandsunderraman opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow our contribution guidelines
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**

This is an initial commit to create a repository that generates go lang based code from an async api spec.
This PR is meant to replace the existing repository.
I did research on the existing repository and tried to make it work and make sense of it.
However after 2 weeks of research it seemed futile. Hence after discussing the same with @derberg it was decided to go with a re-write using a react based template

The current PR uses a go lang event based middleware called [watermill](https://github.com/ThreeDotsLabs/watermill/)
This initial commit will generate a subscriber in go lang for AMQP protocol alone.
The goal is to get initial feedback and then build on this repository

smoya commented on components/Handlers.js:
```diff
@@ -0,0 +1,47 @@
+
+
+let subscriptionFunction = (channelName, operation, message) => `
+// subscription handler for ${channelName}        
```
Go doc tells you documentation for functions should start with the name of the function. For example:

```go
// OnLightMeasured is the subscription handler for ${channelName}.
func OnLightMeasured (msg *message.Message) error {}
```

This applies to all the comments you wrote on the generated code. Also to any type, not only functions.

smoya commented on components/Router.js:
```diff
@@ -0,0 +1,57 @@
+	"${moduleName}/handlers"
+)
+
+func GetRouter() (*message.Router, error){
```
Missing documentation here. `// GetRouter ...`

smoya commented on components/Router.js:
```diff
@@ -0,0 +1,57 @@
+)
+
+func GetRouter() (*message.Router, error){
+
```
nitpick: unnecessary empty line.

smoya commented on components/Router.js:
```diff
@@ -0,0 +1,57 @@
+func GetRouter() (*message.Router, error){
+
+	logger := watermill.NewStdLogger(false, false)
+	router, err := message.NewRouter(message.RouterConfig{}, logger)
```
you can simplify this code by just doing `return message.NewRouter(message.RouterConfig{}, logger)` here

smoya commented on components/Router.js:
```diff
@@ -0,0 +1,57 @@
+  let amqpRules = ""
+  if (subscriberFlags.hasAMQPSub) {
+    amqpRules = `
+func ConfigureAMQPSubscriptionHandlers(r *message.Router, s message.Subscriber) {
```
Missing documentation here. `// ConfigureAMQPSubscriptionHandlers ...`

smoya commented on components/Subscriber.js:
```diff
@@ -0,0 +1,49 @@
+func GetAMQPSubscriber(amqpURI string) (*amqp.Subscriber, error) {
+    amqpConfig := amqp.NewDurableQueueConfig(amqpURI)
+
+    amqpSubscriber, err := amqp.NewSubscriber(
```
You can simplify this code by just doing: `return amqp.NewSubscriber(...)`

smoya commented on components/Subscriber.js:
```diff
@@ -0,0 +1,49 @@
+//render an AMQP subscriber
+function AMQPSubscriber() {
+    return `
+func GetAMQPSubscriber(amqpURI string) (*amqp.Subscriber, error) {
```
Missing documentation here. `// GetAMQPSubscriber ...`

smoya commented on readme.md:
````diff
@@ -1,161 +1,172 @@
-// Populates p with channel parameters
-// p = { StreetlightId: "<my-id>" }
-```
+1. go 1.16 +
````
However, you require version 1.15 in `go.mod` file.

smoya commented on components/Server.js:
```diff
@@ -0,0 +1,27 @@
+function AMQPURI() {
+    return `
+func GetAMQPURI() string {
```
Missing documentation here. // GetAMQPURI ...

smoya commented on template/go.mod.js:
```diff
@@ -0,0 +1,36 @@
+  return (
+    <File name="go.mod">
+  {`
+module ${params.moduleName}
```
What do you think about asking the user to provide a package name?

anandsunderraman replied:
So currently this is an option that the use can pass to the template so I am guessing you were hinting at that ?

smoya replied:
Yep 👍

smoya commented on template/index.js:
```diff
@@ -0,0 +1,92 @@
+
+  ${subscriberConfig}
+
+  ctx := context.Background()
```
What do you think about making the context aware of OS signals so the context gets canceled when the app is interrupted?

```suggestion
  ctx := context.Background()
  ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM, syscall.SIGINT)
  defer stop()
```

anandsunderraman replied:
i am going to include this change which however would mean that we support go 1.16 and upwards only

smoya reviewed (changes requested):
NOTE: I only reviewed generated Go code.

Nice job 👏 👏 
I dropped some comments and request changes on my review.

I would also like to suggest you unify all the different go packages into just one, as they share the same concerns. Meaning `config`, `handlers`, etc in to just one.

smoya commented on readme.md:
```diff
@@ -1,161 +1,172 @@
 
-## What is it?
-AsyncAPI To-Go is a code generation template for the [AsyncAPI generator](https://github.com/asyncapi/generator) that can be used to generate boilerplate Go code for interacting with event-driven architectures. 
+#### Run the following command to generate a golang module
```
```suggestion
#### Run the following command to generate a Go module
```

smoya commented on readme.md:
```diff
@@ -1,161 +1,172 @@
 
-## What Am I Getting Myself Into? 
-Code Samples:
+The above code currently generates a golang module that has a AMQP subscriber.
```
```suggestion
The above code currently generates a Go module that has a AMQP subscriber.
```

smoya commented on readme.md:
````diff
@@ -1,161 +1,172 @@
-// p = { StreetlightId: "<my-id>" }
-```
+1. go 1.16 +
+2. rabbitmq-server
````
Sorry, this is wrong. You should run `docker run -d -p 15672:15672 -p 5672:5672 rabbitmq:3-management` if you wanna also have the UI running at `localhost:15672` for creating queue, publishing message, etc.

anandsunderraman commented on components/Handlers.js:
```diff
@@ -0,0 +1,47 @@
+    var lm payloads.${message}
+    err := json.Unmarshal(msg.Payload, &lm)
+    if err != nil {
+        fmt.Printf("error unmarshalling message: %s, err is: %s", msg.Payload, err)
```
hmm @smoya why do you say infinite loop. should I allow this to return an error and the function invoking it should decide if they should nack it or dead letter it ?

smoya replied:
From what I've tested, once you produce a message that is not JSON, the consumer gets stuck reading the same message again and again. I think this is due to the fact you are returning an error, and as far as I read here https://watermill.io/docs/messages-router/#ack, ACK of the message is not happening.

smoya replied:
Perhaps logging the json unmarshal error is enough, in particular for this generated code.

smoya commented on template/index.js:
```diff
@@ -0,0 +1,95 @@
+  router, err := config.GetRouter()
+  if err != nil {
+    log.Fatalf("error creating watermill router: %s", err)
+    return
```
You don't need the return here. `log.Fatal` does a `panic` call, which invokes an `os.Exit(1)` at the end.
```suggestion
```

Same applies to other parts of the code that use `log.Fatal`

smoya commented on components/Handlers.js:
```diff
@@ -0,0 +1,45 @@
+    err := json.Unmarshal(msg.Payload, &lm)
+    if err != nil {
+        log.Fatalf("error unmarshalling message: %s, err is: %s", msg.Payload, err)
+        return err
```
This is not gonna be executed ever if you run a `log.fatal` before. `log.Fatal` terminates the execution of the program.

smoya commented on components/Handlers.js:
```diff
@@ -0,0 +1,45 @@
+
+    var lm payloads.${message}
+    err := json.Unmarshal(msg.Payload, &lm)
+    if err != nil {
```
From what I've tested, once you produce a message that is not JSON, this `err` will be returned and the consumer gets stuck reading the same message again and again. I think this is due to the fact you are returning an error, and as far as I read here https://watermill.io/docs/messages-router/#ack, ACK of the message is not happening.

Perhaps logging the json unmarshal error is enough, in particular for this generated code.

anandsunderraman replied:
yes makes sense addressed it

smoya commented on components/Router.js:
```diff
@@ -0,0 +1,54 @@
+  let amqpRules = ""
+  if (subscriberFlags.hasAMQPSub) {
+    amqpRules = `
+//ConfigureAMQPSubscriptionHandlers configures the router with the subscription handler    
```
There is an small but important fix to do here. **All** documented types in go should have an space at the beginning of the comment. A minor also is adding a final period. All is by convention. 
```suggestion
// ConfigureAMQPSubscriptionHandlers configures the router with the subscription handler.    
```

Same applies to all comments your wrote.

anandsunderraman commented on readme.md:
````diff
@@ -1,161 +1,172 @@
+2. Run the following commands to download the dependencies
+```bash
+go mod download
+go mod tidy
````
if i do not run `go mod tidy` i get an error like

```
missing go.sum entry for module providing package github.com/ThreeDotsLabs/watermill (imported by go-amqp/asyncapi); to add:
	go get go-amqp/asyncapi
asyncapi/subscribers.go:6:5: missing go.sum entry for module providing package github.com/ThreeDotsLabs/watermill-amqp/pkg/amqp (imported by go-amqp/asyncapi); to add:
	go get go-amqp/asyncapi
asyncapi/handlers.go:9:2: missing go.sum entry for module providing package github.com/ThreeDotsLabs/watermill/message (imported by go-amqp/asyncapi); to add:
	go get go-amqp/asyncapi
```

smoya replied:
WDYT about generating the `go.sum` file as well so all checksums are already checked in?

anandsunderraman replied:
since this code is being generated, it would not be an easy task to generate the go.sum
Also from a tooling perspective they way I see it is that this tool is responsible for generating the required code and basic skeleton. From there on it is the responsibility of the consumer to download the required dependencies

smoya replied:
`go.sum` stores the sum (hash) of all dependencies for consistency check. It is not storing dependencies or anything else. 
Generating the `go.sum` should be an atomic process, so it is as easy as just generating a file containing exactly the same content of your local `go.sum` file.

Then, consumers will need only to do `go mod download` for getting an **exact** copy of what you want.

anandsunderraman replied:
hmm I see what you say, however I have 2 follow up questions

1. In future say I want to add a new dependency to the template. Then I will have to add the new dependency in the template and then I would have to generate the code , run `go mod download` so that the `go.sum` file is generated. Is there a good way to automate this process ?
2. Currently we allow users to pass the `goVersion` as a parameter to this template, will the `go.sum` work across all version of `go`

smoya replied:
> 1. In future say I want to add a new dependency to the template. Then I will have to add the new dependency in the template and then I would have to generate the code , run `go mod download` so that the `go.sum` file is generated. Is there a good way to automate this process ?

I feel is the natural process to follow. Otherwise, how would you add a dependency to Go generated code without ensuring that generated code can, let's say, compile?
You could perhaps add a end-to-end functional test where you generate code and run testsuite, linters if you want, etc.
I think better to open a new issue in order to track this effort.

> 2. Currently we allow users to pass the `goVersion` as a parameter to this template, will the `go.sum` work across all version of `go`

Whatever version you add as directive to the `go.mod` file indicates the minimum acceptable Go version by that module. Meaning users should not enter a lower version than that.

anandsunderraman replied:
@smoya then should I not provide the `goVersion` option to the generator and say that users must use go 1.16+ in the readme ?

smoya replied:
`goVersion` option is used **only** to add the constraint to the `go.mod` file. However, that constraint is tied to the generated code, so I think I would +1 to remove it. Maybe if you can validate the `goVersion` option to be 1.16 or greater... but still IMHO is not worth it.

anandsunderraman replied:
this has now been addressed

anandsunderraman commented:
@smoya I have addressed all your comments with my last push

smoya commented on components/Handlers.js:
```diff
@@ -0,0 +1,44 @@
+    var lm ${message}
+    err := json.Unmarshal(msg.Payload, &lm)
+    if err != nil {
+        fmt.Printf("error unmarshalling message: %s, err is: %s", msg.Payload, err)
```
Shouldn't this be using `log` instead of `fmt` ?

anandsunderraman replied:
you are right fixing this

derberg commented on readme.md:
````diff
@@ -1,161 +1,181 @@
+```bash
+npm install -g @asyncapi/generator
+# clone this repository and navigate to this repository
+ag /path/to/asyncapi.yaml ./ -o /path/to/generated-code -p moduleName=your-go-module-name -p goVersion=1.17
````
```suggestion
ag /path/to/asyncapi.yaml @asyncapi/go-watermill-template -o /path/to/generated-code -p moduleName=your-go-module-name -p goVersion=1.17
```

we will release it under official name

derberg commented on package.json:
```diff
@@ -1,24 +1,110 @@
-  "name": "go-test",
-  "version": "1.0.0",
-  "description": "",
+  "name": "go-template",
```
```suggestion
  "name": "@asyncapi/go-watermill-template",
```

derberg commented on package.json:
```diff
@@ -1,24 +1,110 @@
+    "testTemplate": "rimraf test/temp && jest --coverage -t 'templateGenerationResult'",
+    "testHooks": "rimraf test/temp && jest --coverage -t 'generateExtraFormats'",
+    "testFilters": "rimraf test/temp && jest --coverage -t 'generateMermaidDiagram'",
+    "release": "echo 'We do not trigger releases for this template. Normally this script should invoke \"semantic-release\"'",
```
this time we want to release it

```suggestion
    "release": "semantic-release",
```

derberg commented on package.json:
```diff
@@ -1,24 +1,110 @@
+    "lint": "eslint --max-warnings 0 --fix --config .eslintrc .",
+    "generate:readme:toc": "markdown-toc -i README.md",
+    "generate:assets": "npm run generate:readme:toc",
+    "bump:version": "echo 'We do not trigger releases for this template. Normally this script should invoke \"npm --no-git-tag-version --allow-same-version version $VERSION\"'"
```
```suggestion
    "bump:version": "npm --no-git-tag-version --allow-same-version version $VERSION"
```

derberg commented on readme.md:
````diff
@@ -1,161 +1,181 @@
+
+As of this initial commit this template has been tested to generate an AMQP subscriber for the following asyncapi.yml file
+
+```yaml
````
maybe better put this file in a real file, in test folder, and then your instructions in next sections, to run generator, can point to this file?

derberg commented on readme.md:
```diff
@@ -1,161 +1,181 @@
 
-## What Am I Getting Myself Into? 
-Code Samples:
+1. moduleName: name of the go module to be generated
```
It should be in section like https://github.com/asyncapi/nodejs-template#template-configuration that you already have mentioned in the TOC

derberg commented on readme.md:
````diff
@@ -1,161 +1,181 @@
+   ```
+   docker run -d -p 15672:15672 -p 5672:5672 rabbitmq:3-management
+   ```
+6. Create a queue as per the async api spec
````
```suggestion
6. Create a queue as per the AsyncAPI file
```

derberg replied:
honestly would be cool to have at least a link to doc that explains how to do it

anandsunderraman replied:
the official docs are insufficient, so I am going to provide a link to an article that walks through the UI process.

derberg commented on readme.md:
````diff
@@ -1,161 +1,181 @@
+   docker run -d -p 15672:15672 -p 5672:5672 rabbitmq:3-management
+   ```
+6. Create a queue as per the async api spec
+7. Publish a message to the queue as per the async api spec
````
```suggestion
7. Publish a message to the queue as per the AsyncAPI file
```

derberg replied:
honestly would be cool to have at least a link to doc that explains how to do it, or simple curl, or something

anandsunderraman replied:
curl is such a good idea I am going to add that

derberg reviewed (commented):
some general comments for improvement

derberg commented on readme.md:
```diff
@@ -1,161 +1,112 @@
+
+### CLI
+
+As of this initial commit this template has been tested to generate an AMQP subscriber for the asyncapi.yaml file located in the `test` folder
```
```suggestion
This template has been tested to generate an AMQP subscriber for [this asyncapi.yaml file](./test/asyncapi.yaml).
```

how about this? we link to the file and also do not refer to a commit but the template in generate, at state when people read the readme file

anandsunderraman commented:
@smoya and @derberg I have addressed all your comment in this PR.
Once you review this PR, I would like to close this in lieu of https://github.com/asyncapi/go-template/pull/25
The reason is that this PR did not branch of the master of this repo and hence is not the correct one to merge.This PR was branched of https://github.com/asyncapi/template-for-generator-templates.

https://github.com/asyncapi/go-template/pull/25 is a branch of the current master of this repo and deletes all the unwanted files

derberg commented:
lgtm, I'll added 2 followup questions in PR. Good call!
